### PR TITLE
Particle effects

### PR DESCRIPTION
**Pull request by dslh**
Adds particle effect that emits sparks when a projectile strikes an enemy (see issue #2). I tried to get a screenshot but I'm not fast enough.

I don't think it's done yet. In particular I'd like to see if I can retrieve the normal at the point of impact and set it as the direction for the emitter, so that sparks erupt from the surface of the asteroid. I have a feeling that will look a bit nicer. But I wanted to get your feedback about a couple of things.

 - Where do you think I should be attaching emitters to the scene tree? I'm pretty sure that `get_tree().get_root().add_child()` is a bad habit. Maybe I should add a node to the game scene for this purpose? How would I access it? Or should the emitter be a child of the asteroid/enemy? Or of the projectile? But then how would I prolong the parent's life until the emitter has finished? How do I make sure the emitted sparks are masked by the game area's border?
 - It feels a bit awkward to add a timer to the game scene to clean up these particle effects. Someone suggested it in a forum but I'm not sure how much I should trust an internet stranger. Would be nice to have an event that I can hook into when the emitter finishes emitting but it seems there is no such thing. Another suggestion is to create a class that manages a pool of emitters, so that they can be reused once they're done emitting 🤔 

Looking forward to your thoughts 🙂

### CODE REVIEW

**Comment by JWR26** [comment]
OK. After looking over the code I would have done exactly the same. However, seeing it from the "outside" I can see two "issues" created by this. 

1. I would have attached it to the scene tree just like you did. I have always done it like that. Each particle effect had its own timer and then deleted itself after a set time. I am now wondering if an Autoload on the lines of `EffectContainer` that is a just a simple timer that will delete all non-emmitting children as you suggest. It would gather all of the clutter in one place. However, we would need to consider lots of things to ensure the particle effets are drawn in the right place, on the correct layer etc. While thinking about it, it may seem like a headache for now, but could lead to a nice solution if projects were to scale in size. Time spent now = time saved later mentality.

2. For the particles on rigid bodies, I think we should be creating them from the `_integrate_forces(state: Physics2DDirectBodyState)` method, checking the physics body state for a collision. This would give us the information you would like such as collision normal, body velocity etc, that could then be given to the particle effect to create some really nice, fluid particles that give the feeling of projectiles "smashing to bits". 

What are your thoughts?

On another note, i have a new branch open where I am refactoring the different entities to address issue #4. I'd really appreciate if I could put you as "reviewer" on the branch to make sure I have not removed any features and that the structure I have put in place seems good to you. Hopefully it will give us a good base for extra fun!

**Comment by dslh** [comment]
Thanks for the feedback! I've been experimenting a bit with `_integrate_forces`, it's nothing that's ready to merge but I've pushed a commit to this PR so that you can take a look.

Using `_integrate_forces` made me wonder, which entity is responsible for spawning the emitter? The asteroid or the projectile? Maybe it's both, the projectile emits sparks and the asteroid emits chunks of rock.

I'm finding the iterative development loop of "make a change -> start the game -> try to observe the change while also playing the game -> quit out and make more changes" isn't very productive. I think I need a test rig, like maybe a scene that fires a single projectile at a static (or slowly moving) asteroid. Have you ever set up something like that? There's quite a few adjustments I'd like to try for the particle emitter, it would be nice to speed up the process a bit.

The question of where to add particle effects to the tree is still an open one for me, my latest code calls `get_parent().add_child()` which seems a little bit better. Let's see how it goes as we add more effects, maybe a pattern will emerge that'll guide us to better standards and practises.

**Comment by dslh** [comment]
I'd be very happy to take a look at your branch. I'd suggest opening a PR as a place for discussion.

**Comment by JWR26** [comment]
Firstly, and most importantly, I hope you are having a nice time over in NZ!

I have looked over your changes, and here are my thoughts:

- `test/rigs` is **amazing**. The idea of it, how you have implemented it, the use of `editor discription` to give information about the tests, are all great. Also the two tests you have put really make sense! I'd say this is an excellent practice to take forward.
- I like the use of groups to free the explosions in the scene tree once a single timer has expired in the  `game.gd`. In my oppinion this is a good way to keep the scene tree uncluttered.

For the `bullet_sparks.tscn`:
- I'm not sure I would have the asteroid spawning the bullet particles. I would have thought that each `entity` would be responsible for its own particles. However, I have tried getting this working locally, and for some reason I can't get this to work.
- I wonder if setting the `direction` property to the normal of the collision could be used?
- Perhaps a the colour could be set from code to match that of the projectile. In #10 the projectiles colour is now set from code, so perhaps we could cascade this onto `bullet_sparks.tscn`. Perhaps a script is required? On the subject of colour, how about a gradient so the individual particles fade to transparent before dissapearing? perhaps the `color_initial_ramp` property of the `ParticleMaterials` could be used for this.

**Comment by dslh** [comment]
Thanks for your feedback! I updated the branch so that projectiles and asteroids each release their own particles when they are struck, the style could definitely be improved but at least the combination of two particle systems works well in my opinion.


> I wonder if setting the direction property to the normal of the collision could be used?

Makes sense, I've tried this and it works well. It makes me a little bit nervous because `direction` is a property of the `ParticlesMaterial` class rather than the `Particles2D` node itself. It seems like a single instance of the material class should be shared between particle emitters, so changing properties on it would cause weirdness. But I experimented and it seems to be ok, so I guess there is one instance created for each emitter 🤷 

> Perhaps a the colour could be set from code to match that of the projectile.

Sure, I'll rebase this branch on top of #10 and try to make it fit. Gradients is a good idea, I'll play around with that.

File: effects/sparks.gd [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class_name Sparks
+
+static func emit_from_collisions(state, sparks, container):
+	for i in state.get_contact_count():
+		var normal = state.get_contact_local_normal(i)
+
+		var emitter = sparks.instance()
+		emitter.position = state.get_contact_collider_position(i)
+		emitter.process_material.direction = Vector3(normal.x, normal.y, 0)
+		container.add_child(emitter)
```
Comment (JWR26) [inline effects/sparks.gd]: I really like how you create the sparks. I'll make a mental note of using static functions in classes from now on.

File: entities/projectile.gd [hunk 2]
```diff
@@ -33,10 +33,16 @@ func set_can_bounce(value: bool = false) -> void:
 		$Polygon2D.set_color("#ff8426") 
 		bounce_number = 3
 
+func emit_sparks() -> void:
+	var emitter = sparks.instance()
+	emitter.position = self.position
+	get_tree().get_root().add_child(emitter)
+
```
Comment (JWR26) [inline entities/projectile.gd]: This will need to go into the `projectile2d.gd` file. I will delete this file from the main branch after we merge this PR.

**Review by JWR26 — COMMENTED** [review]
The logic added to projectle.gd can be moved into projectile2d.gd from the main branch and I think we are good to merge!

**Comment by JWR26** [comment]
Awesome retro look. Love it!

**Pull request merged**